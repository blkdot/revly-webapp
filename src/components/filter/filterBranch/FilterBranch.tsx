import { useClickAwayListener, useVendors } from 'hooks';
import { ButtonKit, RadioKit } from 'kits';
import { useRef, useState, useMemo, type ReactNode } from 'react';
import { FaChevronDown, FaChevronRight } from 'react-icons/fa';
import { useAutoAnimate } from '@formkit/auto-animate/react';
import { TChainData } from 'hooks/useVendors';
import List from '@mui/material/List';
import { ListItemText } from '@mui/material';
import ExpandLess from '@mui/icons-material/ExpandLess';
import ExpandMore from '@mui/icons-material/ExpandMore';
import './FilterBranch.scss';

const FilterBranch: React.FC<{
  items: TChainData[];
  values: string[];
  icon?: ReactNode;
  onChange: (k: string) => void;
  label: string;
}> = ({ items, values, onChange, label, icon }) => {
  const [isOpen, setIsOpen] = useState(false);
  const [openAccordion, setOpenAccordion] = useState('');
  const { vendors } = useVendors();
  const chains = useMemo(
    () => Object.keys(vendors.display).sort((a) => (a.trim() === '' ? 1 : -1)),
    [vendors.display]
  );

  const refDropdown = useRef(null);
  const [regfItemDropdown] = useAutoAnimate();
  const [regfDropdown] = useAutoAnimate();

  const selectItem = (v) => {
    onChange(v);
    setIsOpen(false);
  };

  const getCurrentValue = () => {
    const lengthValues = values.length;
    const max = 1;

    if (lengthValues < 1) {
      return label;
    }

    if (lengthValues === items.length) {
      if (items.length === 1)
        return items.find((item) => values[0] === item.vendor_id)?.vendor_name;

      return `All ${label} selected`;
    }

    if (lengthValues > max) return `${lengthValues} ${label} selected`;

    return items
      .filter((item) => values.includes(String(item.vendor_id)))
      .map((val) => val.vendor_name)
      .join(', ');
  };

  useClickAwayListener(refDropdown, () => setIsOpen(false));

  const renderItem = (branches: TChainData[]) =>
    branches.map((item) => (
      <div
        className={`comp-dropdown__item ${
          values.includes(String(item.vendor_id)) ? '__active' : ''
        }`}
        key={String(item.vendor_id)}
        onClick={() => selectItem(String(item.vendor_id))}
        onKeyDown={() => selectItem(String(item.vendor_id))}
        role='button'
        tabIndex={0}
        style={{ width: '100%' }}
      >
        <RadioKit
          checked={values.includes(String(item.vendor_id))}
          onChange={() => selectItem(String(item.vendor_id))}
        />
        <span>{String(item.vendor_name)}</span>
      </div>
    ));

  const renderNewItem = () =>
    chains.map((name) => {
      if (!name) {
        return (
          <div key={`${name}_key_empty`} role='button' tabIndex={0} style={{ width: '100%' }}>
            {renderItem(items.filter((item) => item.chain_name === name))}
          </div>
        );
      }

      return (
        <div
          key={name}
          onClick={() => setOpenAccordion((prev) => (prev === name ? '' : name))}
          onKeyDown={() => setOpenAccordion((prev) => (prev === name ? '' : name))}
          role='button'
          tabIndex={0}
          style={{ cursor: 'pointer', width: '100%' }}
          ref={regfItemDropdown}
        >
          <ListItemText style={{ width: '100%' }}>
            <div
              style={{
                display: 'flex',
                justifyContent: 'space-between',
                alignItems: 'center',
                width: '100%',
                marginTop: '0.6rem',
              }}
            >
              {name}
              {openAccordion === name ? <ExpandLess /> : <ExpandMore />}
            </div>
          </ListItemText>
          {openAccordion === name ? (
            <div key={name} style={{ width: '100%' }}>
              {renderItem(items.filter((item) => item.chain_name === name))}
            </div>
          ) : null}
        </div>
      );
    });

  const renderItems = () => {
    if (!isOpen) return null;

    return (
      <div className='comp-dropdown__content'>
        {renderNewItem()}
      </div>
    );
  };

  const renderIcon = () => {
    if (!icon) return null;

    return (
      <i
        style={{ marginRight: '0.5rem', alignItems: 'center' }}
        className={`${values.length > 0 ? '__active' : ''}`}
      >
        {icon}
      </i>
    );
  };

  return (
    <div className='comp-dropdown__branch' ref={refDropdown}>
      <ButtonKit
        variant='outlined'
        onClick={(e) => {
          e.stopPropagation();
          setIsOpen((prev) => !prev);
        }}
        role='button'
        style={{ display: 'flex', justifyContent: 'space-between' }}
        className={`${values.length > 0 ? '__active' : ''}`}
        tabIndex={0}
      >
        <div className={`comp-content ${values.length > 0 ? 'active' : ''}`}>
          {renderIcon()}
          {getCurrentValue()}
        </div>
        <div
          className='comp-dropdown-arrow'
          style={{ display: 'flex', marginLeft: '1rem', alignItems: 'center' }}
        >
          {isOpen ? <FaChevronRight /> : <FaChevronDown />}
        </div>
      </ButtonKit>
      <List component='nav' style={{ display: 'flex', flexDirection: 'column' }} ref={regfDropdown}>
        {renderItems()}
      </List>
    </div>
  );
};

export default FilterBranch;
