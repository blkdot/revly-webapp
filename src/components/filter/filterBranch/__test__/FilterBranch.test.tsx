import renderer from 'react-test-renderer';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import FilterBranch from '../FilterBranch';

jest.mock('axios');
jest.mock('@formkit/auto-animate/react');
jest.mock('hooks/usePlatform');

const items = [
  {
    chain_name: "Adam's Kitchen",
    chain_id: 82369,
    vendor_id: '126601',
    vendor_name: "Adam's Kitchen",
    is_active: true,
    platform: 'deliveroo',
    access_token: '',
    access_token_bis: null,
    data: {
      vendor_id: '126601',
      chain_id: 82369,
      data: {
        vendor_name: "Adam's Kitchen",
        chain_name: "Adam's Kitchen",
      },
      metadata: {
        cost: '0.15',
        drn_id: '79781475-****-****-****-****07b',
        org_id: '0e35a3d3-****-****-****-****a24',
        is_active: true,
        is_deleted: false,
      },
    },
  },
];

test('Should render Branch Filter', () => {
  const queryClient = new QueryClient();

  const mockOnchange = jest.fn();

  const component = renderer.create(
    <QueryClientProvider client={queryClient}>
      <FilterBranch label='test' values={['value1']} onChange={mockOnchange} items={items} />
    </QueryClientProvider>
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
