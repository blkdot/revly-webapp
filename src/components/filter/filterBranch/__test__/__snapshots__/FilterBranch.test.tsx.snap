// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Branch Filter 1`] = `
<div
  className="comp-dropdown__branch disabled"
>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium __active css-8xewcm-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="button"
    style={
      Object {
        "display": "flex",
        "justifyContent": "space-between",
      }
    }
    tabIndex={0}
    type="button"
  >
    <div
      className="comp-content active"
    />
    <div
      className="comp-dropdown-arrow"
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "marginLeft": "1rem",
        }
      }
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 448 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
        />
      </svg>
    </div>
  </button>
  <nav
    className="MuiList-root MuiList-padding css-h4y409-MuiList-root"
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
      }
    }
  />
</div>
`;
